File: src/main/java/logic/CollectibleObject.java
package logic;

import lib.AudioUtility;

public /* fill code */ class CollectibleObject /* fill code */{
	
	protected int requiredGrabbingTime;
	protected int grabbingTimeCount;
	
	public CollectibleObject(int radius, int movingDuration, int z, int requiredGrabbingTime) {
		/* fill code */
	}
	
	public void ungrab(){
		/* fill code */
	}

	public void grab(PlayerStatus player){
		if(isDestroyed) return;
		if(/* fill code */){
			AudioUtility.playSound("collect");
			collect(player);
			isDestroyed = true;
			return;
		}
		/* fill code */
	}
	
	public abstract void collect(PlayerStatus player);
}

File: src/logic/CollectibleObject.kt
package logic

import lib.AudioUtility

abstract class CollectibleObject(radius: Int, movingDuration: Int, z: Int, private val requiredGrabbingTime: Int) : TargetObject(radius, movingDuration, z) {
    private var grabbingTimeCount = 0

    fun ungrab() {
        grabbingTimeCount = 0
    }

    fun grab(player: PlayerStatus) {
        if (isDestroyed)
            return
        if (grabbingTimeCount >= requiredGrabbingTime) {
            AudioUtility.playSound("collect")
            collect(player)
            isDestroyed = true
            return
        }
        grabbingTimeCount++
    }

    abstract fun collect(player: PlayerStatus)
}
